// react-maps/lib/info-window.js
/* global google */

'use strict';

var React = require('react');
var ReactDOMServer = require('react-dom/server');
var blacklist = require('blacklist');

module.exports = React.createClass({
  displayName: 'exports',

  getDefaultProps: function getDefaultProps() {
    return {
      // WARNING: if autoPan is enabled, the parent maps props will be overridden
      disableAutoPan: true
    };
  },

  componentWillUnmount: function componentWillUnmount() {
    if (!this.iw) return;

    this.iw.close();
  },

  render: function render() {
    var props = this.props;
    var options = blacklist(this.props, 'anchor', 'children', 'map', 'open');

    if (props.children) {
      options.content = ReactDOMServer.renderToStaticMarkup(React.createElement('span', {}, props.children));
    }

    // new
    if (!this.iw) {
      this.iw = new google.maps.InfoWindow(options);

      // existing
    } else {
        this.iw.setOptions(options);
      }

    if (props.open) {
      if (!props.map) return null;

      this.iw.open(props.map, props.anchor);
    } else {
      this.iw.close();
    }

    return null;
  }
});
